crates/core/src/task_center.rs

```diff
@@ -625,16 +628,22 @@ impl TaskCenterInner {
             })
             .unwrap();
 
-        Ok(RuntimeRootTaskHandle {
-            inner_handle: result_rx,
-            cancellation_token: cancel,
-        })
+        Ok(RuntimeTaskHandle::new(runtime_name, cancel, result_rx))
     }
 
+    /// Runs **only** after the inner main thread has completed work and no other owner exists for
+    /// the runtime handle.
     fn drop_runtime(self: &Arc<Self>, name: &'static str) {
         let mut runtimes_guard = self.managed_runtimes.lock();
-        if runtimes_guard.remove(name).is_some() {
-            trace!("Runtime {} was dropped", name);
+        if let Some(runtime) = runtimes_guard.remove(name) {
+            // We must be the only owner of runtime at this point.
+            let name = runtime.name().to_owned();
+            debug!("Runtime {} completed", runtime.name());
+            let owner = Arc::into_inner(runtime.into_inner());
+            if let Some(runtime) = owner {
+                runtime.shutdown_timeout(Duration::from_secs(2));
+                trace!("Runtime {} shutdown completed", name);
+            }
```

Comment:
Should we add an expect to encode the invariant that we are the only owner of the runtime at this point?

Reply:
I'm not sure as it wouldn't impact correctness of the system. Let's add it when we face a scenario where we need to really be sure that we uphold it.